### PR TITLE
Flatpacks and the Flatpacker 1001

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

Formalizes the existence of flatpacks, small boxes that can be unpacked into machines via a multitool. These already existed for solar panels, station beacons, and AME parts, but now they can be created for any machine. This is done via the Flatpacker 1001, which, using the machine board inserted inside of it and some materials, can create a flatpack. 

The cost of the flatpack is based on the material costs of the components it takes to assemble the machine + a small extra tax. Assembling single machines via flatpacks isn't cost efficient, but assembling a large number of them is, since you don't need to reprint machine boards.

## Why / Balance

Explained at length [in the machine upgrading rework doc.](https://docs.spacestation14.com/en/proposals/emogarbage-machine-upgrading-rework.html)

## Technical details

AMEPartSystem and its component have been fully scrapped in favor of the new flatpack system.

## Media

image
image
image
image
image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Added flatpacks, boxes that unpack into a full-sized machine when activated with a multitool.
- add: Added the Flatpacker 1001 to industrial research. A machine capable of creating flatpacks from machine boards.

### CODE REVIEW

**Comment by Aisu9** [comment]
thanks god setting shop with this bad boy will be so fast and easy

**Comment by CrigCrag** [comment]
this is rather good

**Comment by Hmeister-fake** [comment]
This should have some sort of Emag interaction that lets you flatpack living and dead entities, would make stealing animals more easy as a thief.

**Comment by EmoGarbage404** [comment]
> This should have some sort of Emag interaction that lets you flatpack living and dead entities, would make stealing animals more easy as a thief.

would require fundamentally rewriting the system + it's literally not how the object works in the in-universe logic. The board is never consumed.

File: Content.Client/Construction/UI/FlatpackCreatorMenu.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    public override void Close()
+    {
+        base.Close();
+
+        _entityManager.DeleteEntity(_machinePreview);
+        _machinePreview = null;
+    }
+}
```
Comment (metalgearsloth) [inline Content.Client/Construction/UI/FlatpackCreatorMenu.xaml.cs:141]: If I keep the BUI open into lobby this doesn't cause issues does it.

File: Content.Server/Construction/FlatpackSystem.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            return;
+
+        if (!TryComp<MachineBoardComponent>(machineBoard, out var boardComp))
+            return;
+
+        if (!MaterialStorage.CanChangeMaterialAmount(uid, GetFlatpackCreationCost(ent, (machineBoard, boardComp))))
+            return;
+
+        comp.Packing = true;
+        comp.PackEndTime = _timing.CurTime + comp.PackDuration;
```
Comment (metalgearsloth) [inline Content.Server/Construction/FlatpackSystem.cs:43]: Add pause time

File: Content.Shared/Construction/SharedFlatpackSystem.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            Log.Error($"No entity prototype present for flatpack {ToPrettyString(ent)}.");
+
+            if (_net.IsServer)
+                QueueDel(ent);
+            return;
+        }
+
+        var buildPos = _map.TileIndicesFor(grid, gridComp, xform.Coordinates);
+        var intersecting = _entityLookup.GetEntitiesIntersecting(buildPos.ToEntityCoordinates(grid, _mapManager).Offset(new Vector2(0.5f, 0.5f)),
+            LookupFlags.Dynamic | LookupFlags.Static);
```
Comment (metalgearsloth) [inline Content.Shared/Construction/SharedFlatpackSystem.cs:70]: Use the overload that lets you pass in a set instead.

File: Content.Shared/Lathe/SharedLatheSystem.cs [hunk 4]
```diff
@@ -53,4 +58,29 @@ public static int AdjustMaterial(int original, bool reduce, float multiplier)
...
+    }
+
+    private void BuildInverseRecipeDictionary()
+    {
+        _inverseRecipeDictionary.Clear();
+        foreach (var latheRecipe in _proto.EnumeratePrototypes<LatheRecipePrototype>())
+        {
+            _inverseRecipeDictionary.TryAdd(latheRecipe.Result, new List<LatheRecipePrototype>());
+            _inverseRecipeDictionary[latheRecipe.Result].Add(latheRecipe);
+        }
```
Comment (metalgearsloth) [inline Content.Shared/Lathe/SharedLatheSystem.cs:76]: GetOrNew might be better as I don't know if the JIT is smart enough to not construct the list every time.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by K-Dynamic** [comment]
> thanks god setting shop with this bad boy will be so fast and easy

Let's set up shop!

_New construction options_

To clarify, you insert a machine board into the flatpacker as a schematic of sorts, which lets you create a flatpacked version fits in your pocket and be deployed via multitool? And the board isn't used as mentioned prior?

I like this, a bit worried how this impacts gameplay if you have flatpacks to be deployed at a moment's notice

**Comment by UbaserB** [comment]
the more i imagine this idea, the more excited i become.

**Comment by EmoGarbage404** [comment]
> To clarify, you insert a machine board into the flatpacker as a schematic of sorts, which lets you create a flatpacked version fits in your pocket and be deployed via multitool? And the board isn't used as mentioned prior?
 
all flatpacks are size medium and can't fit in pockets, but otherwise yes.

> I like this, a bit worried how this impacts gameplay if you have flatpacks to be deployed at a moment's notice

i mean i guess it could be an instant barrier, but it's a really expensive way of making one. it's significantly cheaper to just make iron foam or something.

File: Content.Shared/Construction/SharedFlatpackSystem.cs (hunk 3 above)
Comment (EmoGarbage404) [reply Content.Shared/Construction/SharedFlatpackSystem.cs]: there wasn't an overload for entitycoordinates / a point that allows passing in a set.

**Review by metalgearsloth — APPROVED** [review]

**Comment by K-Dynamic** [comment]
> i mean i guess it could be an instant barrier, but it's a really expensive way of making one. it's significantly cheaper to just make iron foam or something.

Substations explode on taking damage, I can see an engi trying to suicide bomb with them at least (me lol)

**Pull request merged**